stackslib/src/net/stackerdb/db.rs

```diff
@@ -0,0 +1,554 @@
+        smart_contract: &ContractId,
+        slot_id: u32,
+    ) -> Result<Option<StacksAddress>, net_error> {
+        let stackerdb_id = self.get_stackerdb_id(smart_contract)?;
+        let sql = "SELECT stacker FROM chunks WHERE stackerdb_id = ?1 AND slot_id = ?2";
+        let args: &[&dyn ToSql] = &[&stackerdb_id, &slot_id];
+        query_row(&self.conn, &sql, args).map_err(|e| e.into())
+    }
+
+    /// Get the slot metadata
+    pub fn get_slot_metadata(
+        &self,
+        smart_contract: &ContractId,
+        slot_id: u32,
+    ) -> Result<Option<SlotMetadata>, net_error> {
+        inner_get_slot_metadata(&self.conn, smart_contract, slot_id)
+    }
+
+    /// Get a slot's validation data
+    pub fn get_slot_validation(
+        &self,
+        smart_contract: &ContractId,
+        slot_id: u32,
+    ) -> Result<Option<SlotValidation>, net_error> {
+        inner_get_slot_validation(&self.conn, smart_contract, slot_id)
+    }
+
+    /// Get the list of slot ID versions for a given DB instance at a given reward cycle
+    pub fn get_slot_versions(&self, smart_contract: &ContractId) -> Result<Vec<u32>, net_error> {
+        let stackerdb_id = self.get_stackerdb_id(smart_contract)?;
+        let sql = "SELECT version FROM chunks WHERE stackerdb_id = ?1 ORDER BY slot_id";
+        let args: &[&dyn ToSql] = &[&stackerdb_id];
+        query_rows(&self.conn, &sql, args).map_err(|e| e.into())
+    }
+
+    /// Get the list of slot write timestamps for a given DB instance at a given reward cycle
+    pub fn get_slot_write_timestamps(
+        &self,
+        smart_contract: &ContractId,
+    ) -> Result<Vec<u64>, net_error> {
+        let stackerdb_id = self.get_stackerdb_id(smart_contract)?;
+        let sql = "SELECT write_time FROM chunks WHERE stackerdb_id = ?1 ORDER BY slot_id";
+        let args: &[&dyn ToSql] = &[&stackerdb_id];
+        query_rows(&self.conn, &sql, args).map_err(|e| e.into())
+    }
+
+    /// Get the latest chunk out of the database.
+    /// Returns Ok(Some(data)) if the chunk exists
+    /// Returns Ok(None) if the chunk does not exist
+    /// Returns Err(..) if the DB does not exist, or some other DB error occurs
+    pub fn get_latest_chunk(
+        &self,
+        smart_contract: &ContractId,
+        slot_id: u32,
+    ) -> Result<Option<Vec<u8>>, net_error> {
+        let stackerdb_id = self.get_stackerdb_id(smart_contract)?;
+        let qry = "SELECT data FROM chunks WHERE stackerdb_id = ?1 AND slot_id = ?2";
+        let args: &[&dyn ToSql] = &[&stackerdb_id, &slot_id];
+
+        let mut stmt = self
+            .conn
+            .prepare(&qry)
+            .map_err(|e| net_error::DBError(e.into()))?;
+
+        let mut rows = stmt.query(args).map_err(|e| net_error::DBError(e.into()))?;
+
+        let mut data = None;
+        while let Some(row) = rows.next().map_err(|e| net_error::DBError(e.into()))? {
+            data = Some(row.get_unwrap(0));
+            break;
+        }
+        Ok(data)
```

Comment:
You can get a single row result from a database, wrapped in an option by using `query_row().optional()`. That replaces lines 526-538 with just:

```suggestion
     self.conn.query_row(qry, args, |row| row.get(0)).optional().map_err(|e| e.into())
```